Getter returning Optional<T> of nullable T causing compilation error

Hi!

I'm having some troubles with class containing fields that are nullable and getters for those that return Optional<T>:

```
class Car {
    String someOptionalText;

    public Optional<String> getSomeOptionalText() {
        return Optional.ofNullable(someOptionalText);
    }
}
```

This causes the following error: 

> [ERROR] Line xx: Type mismatch: cannot convert from String to Optional&lt;String&gt;

Looking at the generated Serializer I see that it contains the following:

```
@Override
public Optional<String> getValue(Car bean, JsonSerializationContext ctx) {
    return getValueWithJsni(bean);
}

private native String getValueWithJsni(Car bean) /*-{
    return bean.@the.package.Car::someOptionalText;
}-*/;
```
Which of course isn't working. 

I've used the following config:

```
public class JacksonGWTConfig extends AbstractConfiguration {
    @Override
    protected void configure() {
        fieldVisibility(Visibility.ANY);
        getterVisibility(Visibility.NONE);
        isGetterVisibility(Visibility.NONE);
        setterVisibility(Visibility.NONE);
    }	
}
```
Looking in the code for method parse in PropertyParser I see that it uses the help method parseMethods, which doesn't care about the configuration. So it uses my getter that has a different return type (Optional) and later uses this as the return type for the Serializer.. 

Can this be solved with some config, is this a bug or is there something else I'm missing?

--- a/gwt-jackson/src/main/java/com/github/nmorel/gwtjackson/rebind/property/PropertyProcessor.java
+++ b/gwt-jackson/src/main/java/com/github/nmorel/gwtjackson/rebind/property/PropertyProcessor.java
@@ -204,7 +204,7 @@ public final class PropertyProcessor {
         }
 
         final String propertyName = propertyAccessors.getPropertyName();
-        final JType type = findType( logger, propertyAccessors, typeOracle );
+        final JType type = findType( logger, propertyAccessors, typeOracle, getterAutoDetected, setterAutoDetected, fieldAutoDetected );
 
         PropertyInfoBuilder builder = new PropertyInfoBuilder( propertyName, type );
 
@@ -381,14 +381,14 @@ public final class PropertyProcessor {
         }
     }
 
-    private static JType findType( TreeLogger logger, PropertyAccessors fieldAccessors, JacksonTypeOracle typeOracle ) throws
-            UnableToCompleteException {
+    private static JType findType( TreeLogger logger, PropertyAccessors fieldAccessors, JacksonTypeOracle typeOracle,
+    		boolean getterAutoDetected, boolean setterAutoDetected, boolean fieldAutoDetected ) throws UnableToCompleteException {
         JType type;
-        if ( fieldAccessors.getGetter().isPresent() ) {
+        if ( getterAutoDetected && fieldAccessors.getGetter().isPresent() ) {
             type = fieldAccessors.getGetter().get().getReturnType();
-        } else if ( fieldAccessors.getSetter().isPresent() ) {
+        } else if ( setterAutoDetected && fieldAccessors.getSetter().isPresent() ) {
             type = fieldAccessors.getSetter().get().getParameters()[0].getType();
-        } else if ( fieldAccessors.getField().isPresent() ) {
+        } else if ( fieldAutoDetected && fieldAccessors.getField().isPresent() ) {
             type = fieldAccessors.getField().get().getType();
         } else if ( fieldAccessors.getParameter().isPresent() ) {
             type = fieldAccessors.getParameter().get().getType();

--- a/gwt-jackson/src/test/java/com/github/nmorel/gwtjackson/client/GwtJacksonTestSuite.java
+++ b/gwt-jackson/src/test/java/com/github/nmorel/gwtjackson/client/GwtJacksonTestSuite.java
@@ -102,6 +102,7 @@ import com.github.nmorel.gwtjackson.client.mapper.AllMapsObjectMapperTest;
 import com.github.nmorel.gwtjackson.client.mapper.AutoDetectionGwtTest;
 import com.github.nmorel.gwtjackson.client.mapper.CommonJsonMapperTest;
 import com.github.nmorel.gwtjackson.client.mapper.CustomIterableGwtTest;
+import com.github.nmorel.gwtjackson.client.mapper.DifferentTypeGetterGwtTest;
 import com.github.nmorel.gwtjackson.client.mapper.EnumMapperGwtTest;
 import com.github.nmorel.gwtjackson.client.mapper.ErrorGwtTest;
 import com.github.nmorel.gwtjackson.client.mapper.IgnoreStaticGwtTest;
@@ -287,6 +288,7 @@ public class GwtJacksonTestSuite extends TestCase {
         suite.addTestSuite( ErrorGwtTest.class );
         suite.addTestSuite( AutoDetectionGwtTest.class );
         suite.addTestSuite( EnumMapperGwtTest.class );
+        suite.addTestSuite( DifferentTypeGetterGwtTest.class );
 
         // Annotations test
         suite.addTestSuite( JsonAutoDetectGwtTest.class );

--- a/gwt-jackson/src/test/java/com/github/nmorel/gwtjackson/client/mapper/DifferentTypeGetterGwtTest.java
+++ b/gwt-jackson/src/test/java/com/github/nmorel/gwtjackson/client/mapper/DifferentTypeGetterGwtTest.java
@@ -1,0 +1,20 @@
+package com.github.nmorel.gwtjackson.client.mapper;
+
+import com.github.nmorel.gwtjackson.client.GwtJacksonTestCase;
+import com.github.nmorel.gwtjackson.client.ObjectMapper;
+import com.github.nmorel.gwtjackson.shared.ObjectMapperTester;
+import com.github.nmorel.gwtjackson.shared.mapper.DifferentTypeGetterTester;
+import com.github.nmorel.gwtjackson.shared.mapper.DifferentTypeGetterTester.DifferentTypeGetterBean;
+import com.google.gwt.core.client.GWT;
+
+public class DifferentTypeGetterGwtTest extends GwtJacksonTestCase {
+	
+	public interface DifferentTypeGetterBeanMapper extends ObjectMapper<DifferentTypeGetterBean>, ObjectMapperTester<DifferentTypeGetterBean> {
+        static DifferentTypeGetterBeanMapper INSTANCE = GWT.create( DifferentTypeGetterBeanMapper.class );
+    }
+	
+	public void testDifferentTypeGetter() {
+		DifferentTypeGetterTester.INSTANCE.testDifferentTypeGetterBean(DifferentTypeGetterBeanMapper.INSTANCE);
+	}
+	
+}

--- a/gwt-jackson/src/test/java/com/github/nmorel/gwtjackson/jackson/mapper/DifferentTypeGetterJacksonTest.java
+++ b/gwt-jackson/src/test/java/com/github/nmorel/gwtjackson/jackson/mapper/DifferentTypeGetterJacksonTest.java
@@ -1,0 +1,16 @@
+package com.github.nmorel.gwtjackson.jackson.mapper;
+
+import org.junit.Test;
+
+import com.github.nmorel.gwtjackson.jackson.AbstractJacksonTest;
+import com.github.nmorel.gwtjackson.shared.mapper.DifferentTypeGetterTester;
+import com.github.nmorel.gwtjackson.shared.mapper.DifferentTypeGetterTester.DifferentTypeGetterBean;
+
+public class DifferentTypeGetterJacksonTest extends AbstractJacksonTest {
+
+	@Test
+	public void testDifferentTypeGetter()  {
+		DifferentTypeGetterTester.INSTANCE.testDifferentTypeGetterBean( createMapper(DifferentTypeGetterBean.class) );
+	}
+	
+}

--- a/gwt-jackson/src/test/java/com/github/nmorel/gwtjackson/shared/mapper/DifferentTypeGetterTester.java
+++ b/gwt-jackson/src/test/java/com/github/nmorel/gwtjackson/shared/mapper/DifferentTypeGetterTester.java
@@ -1,0 +1,43 @@
+package com.github.nmorel.gwtjackson.shared.mapper;
+
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility;
+import com.github.nmorel.gwtjackson.shared.AbstractTester;
+import com.github.nmorel.gwtjackson.shared.ObjectMapperTester;
+
+public class DifferentTypeGetterTester extends AbstractTester {
+	
+	// Although the example getter and setter are unlikely,
+	// it is nice to be able to have them use different types than the field.
+	// It allows users to have getters that return Optional<T> of nullable fields of type T.
+	@JsonAutoDetect(fieldVisibility = Visibility.ANY, getterVisibility = Visibility.NONE, setterVisibility = Visibility.NONE)
+	public static class DifferentTypeGetterBean {
+		public boolean fieldWithOtherTypeAccessor;
+		
+		public String getFieldWithOtherTypeAccessor() {
+			return fieldWithOtherTypeAccessor ? "YES" : "NO";
+		}
+		
+		public void setFieldWithOtherTypeAccessor(String fieldWithOtherTypeAccessor) {
+			this.fieldWithOtherTypeAccessor = "YES".equals(fieldWithOtherTypeAccessor);
+		}
+	}
+	
+	public static DifferentTypeGetterTester INSTANCE = new DifferentTypeGetterTester();
+	
+	private DifferentTypeGetterTester() {}
+	
+	public void testDifferentTypeGetterBean(ObjectMapperTester<DifferentTypeGetterBean> mapper) {
+		// This test is actually not that interesting. 
+		// Actual test is the generation of Serializer/Deserializer of DifferentTypeGetterBean.
+		String input = "{\"fieldWithOtherTypeAccessor\":\"true\"}";
+		DifferentTypeGetterBean bean = mapper.read(input);
+		assertEquals( "YES", bean.getFieldWithOtherTypeAccessor() );
+		assertEquals( "{\"fieldWithOtherTypeAccessor\":true}", mapper.write(bean) );
+		
+		bean.setFieldWithOtherTypeAccessor("NO");
+		assertEquals( "NO", bean.getFieldWithOtherTypeAccessor() );
+		assertEquals( "{\"fieldWithOtherTypeAccessor\":false}", mapper.write(bean) );
+	}
+	
+}

